proposals/csharp-9.0/records.md

```diff
@@ -197,6 +197,118 @@ If the containing record is abstract, the synthesized clone method is also abstr
+
+The method:
+1. for each of the record's printable members (public field and property members), appends that member's name followed by " = " followed by the member's value: `this.member`, separated with ", ",
+2. return true if the record has printable members.
+
+If the record type is derived from a base record `Base`, the record includes a synthesized override equivalent to a "print" method declared as follows:
+```C#
+protected override bool print(StringBuilder builder);
+```
+
+If the record has no printable members, the method calls the base "print" method with one argument (its `builder` parameter) and returns the result.
+
+Otherwise, the method:
+1. calls the base "print" method with one argument (its `builder` parameter),
+2. if the "print" method returned true, append ", " to the builder,
+3. for each of the record's printable members, appends that member's name followed by " = " followed by the member's value: `this.member`, separated with ", ",
+4. return true.
+
+The record includes a synthesized method equivalent to a method declared as follows:
+```C#
+public override string ToString();
+```
+
+The method can be declared explicitly. It is an error if the explicit declaration does not match the expected signature or accessibility, or if the explicit declaration doesn't allow overiding it in a derived type and the record type is not `sealed`. It is an error if either synthesized, or explicitly declared method doesn't override `object.ToString()` (for example, due to shadowing in intermediate base types, etc.).
+
+The synthesized method:
+1. creates a `StringBuilder` instance,
+2. appends the record name to the builder, followed by " { ",
+3. invokes the record's "print" method giving it the builder,
+4. appends " }",
+3. returns the builder's contents with `builder.ToString()`.
+
+For example, consider the following record types:
+
+``` csharp
+record R1(T1 P1);
+record R2(T1 P1, T2 P2, T3 P3) : R1(P1);
+```
+
+For those record types, the synthesized printing members would be something like:
+
+```C#
+class R1 : IEquatable<R1>
+{
+    public T1 P1 { get; init; }
+    
+    protected virtual bool print(StringBuilder builder)
+    {
+        builder.Append(nameof(P1));
+        builder.Append(" = ");
+        builder.Append(this.P1);
+        
+        return true;
+    }
+    
+    public override string ToString()
+    {
+        var builder = new StringBuilder();
+        builder.Append(nameof(R1));
+        builder.Append(" { ");
+
+        print(builder);
+
+        builder.Append(" }");
+        return builder.ToString();
+    }
+}
+
+class R2 : R1, IEquatable<R2>
+{
+    public T2 P2 { get; init; }
+    public T3 P3 { get; init; }
+    
+    protected override void print(StringBuilder builder)
+    {
+        if (base.print(builder))
+            builder.Append(", ");
+            
+        builder.Append(nameof(P2));
+        builder.Append(" = ");
+        builder.Append(this.P2);
+        
+        builder.Append(", ");
```

Comment:
Given that this will use the default `ToString` for each member, which typically uses the default culture, should this also use things like [TextInfo.ListSeparator](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.textinfo?view=netcore-3.1) for a culture aware format?

Speaking of globalization, is passing in a culture something that was discussed? The typical variants (as provided on "primitive like types in .NET") are:
```
public override string ToString()
public string ToString(string? format)
public string ToString(IFormatProvider? provider)
public string ToString(string? format, IFormatProvider? provider)
```

Reply:
@tannergooding The idea is to stick closely to `ToString` for anonymous types and tuples. They use comma as separators.
Even though the result of `ToString` is not C# (we're not quoting/escaping strings for instance), it is designed to look like C#. So I think commas make more sense than culture-dependent separators in this scenario.

Reply:
👍, just wanted to raise it as previous reviews by the libraries team have raised concerns over mixing culture sensitive numbers with `,` as it may lead to confusing semantics when read.